### ISSUE TITLE
false positive: subtest (t.Run) where test function is returned from factory

### ISSUE THREAD

**Issue by breml** [kulti/thelper#26]
I have a case, where I do the following:

```Go
func TestIntegration(t *testing.T) {
	tt := []struct {
		name     string
		testFunc func(testFixture *integrationtest.Test) func(*testing.T)
	}{
		{
			name:     "Foo",
			testFunc: testFoo,
		},
	}

	testFixture := integrationtest.New(t)
	defer testFixture.Shutdown()

	for _, tc := range tt {
		t.Run(tc.name, tc.testFunc(testFixture))
	}
}

func testFoo(testFixture *integrationtest.Test) func(*testing.T) {
	return func(t *testing.T) {
		testFixture.Setup()

		// some test logic
	}
}
```

The logic is used to start a database backend as docker container (`integrationtest.New(t)`) and then create a DB with the necessary tables and fixtures (`testFixture.Setup()`) for each test case.

The actual test function, that is executed with `t.Run(...)` is returned from a factory function. In this case, the returned function is not a test helper but the actual test function. But thelper does falsely recognize the returned function as test helper.

Might be related to: #15

**Comment by kulti** [comment]
@breml, thank you for your feedback!

Interesting case… I need to think about to find a way to properly solution.

**Comment by breml** [comment]
@kulti Thanks for your update. I gave it a try and unfortunately, `thelper` (built from master) does still complain for my use-case (mentioned above). I double checked against v0.4.0 and the result is the same. Can you please re-open this issue?

**Comment by kulti** [comment]
This fix was tough and I've forgot to check original case 😂 Sorry for that.

At now, I see, that original case is out of static linter's scope. `tc.testFunc` is a dynamic reference and can be 100% detected only after code compilation.

So, this case won't be fixed.

**Comment by breml** [comment]
OK, I understand, that this is out of scope for a static linter. My question is, if there is still something that we can add to `thelper` to better support this situation.

The use case boils down to the following situation:

* Multiple test cases do need the same initial setup.
* This initial setup is time expensive and therefore should only be done once.
* Each test case is different enough, that it does not make sense to keep all these test cases in a single `Test...(t *testing.T)` function.
* In order to use the initial setup, we need to pass it to the test function. We do this through a factory function, that returns the effective test function.

Initially we added `//nolint:thelper` comments in all locations, but this is not really a good solution, especially with the increasing number of tests.
The second option is to add an exclude rule to `golangci.yml`:

```yaml
issues:
  exclude-rules:
    # Exclude thelper check for t.Helper() in integration tests.
    - path: "_integration_test\\.go$"
      text: "test helper function should start from t.Helper()"
```

But I don't like this solution as well, for two reasons:

1. `thelper` will no longer detect missing `t.Helper()` calls in real helper functions in our integration tests.
2. We depend on the exact error message for the exception.

Do you have some other ideas on how to resolve this issue?

I was thinking a long the lines of having a (maybe configurable) name pattern/prefix for test factory functions where `thelper` then would these functions as if they are regular sub tests (and those not warn about a missing `t.Helper()`.

e.g.

```Go
func testFactoryFoo(testFixture *integrationtest.Test) func(*testing.T) {
	return func(t *testing.T) {
		// missing t.Helper() is OK here, because this function is returned from a `testFactory`
	}
}
```

**Comment by kulti** [comment]
Hi, @breml.

Sorry for the late response.

How about to update your test code in the following way:
```
func TestIntegration(t *testing.T) {
	testFixture := integrationtestNew(t)
	defer testFixture.Shutdown()

	t.Run("Foo", testFoo(testFixture))
	t.Run("Bar", testBar(testFixture))
}
```
The master version of thelper (not released yet) can handle this.

If you can use the existed solutions for test suiting (like testify/suite) it can help too, because test methods have not `*testing.T` parameter. But thelper does not check helper methods too (#29).

I've also thought about using `source` settings inside `exclude-rules` but cannot find a solution to configure goloangci-lint narrowly.

**Comment by breml** [comment]
Hi @kulti 
Thanks for coming back to me with your proposal. I discussed this with my teammates and we agreed to refactor our integration tests the way you suggested. The only thing that is missing right now is, a new release of `thelper` and `golangci-lint`. When do you expect to create a new release?

**Comment by kulti** [comment]
I've thought about another issue to include the both into new release, but have not find a simple way to fix the second.

So, the new release is tagged 🎉
